### PR TITLE
Action button refactor/rework: Enhanced Dragging

### PR DESCRIPTION
**Pull request by SandPoot**
## About The Pull Request
***Ports [tgstation/65180](https://github.com/tgstation/tgstation/pull/65180)***
Check description THERE.

## Why It's Good For The Game
Players have more control over the clutter on their screen.
Buttons are available, but not in the way,
Since any player move of a button saves it, any lack of clarity in the way buttons work will be forced out by buttons not just resetting when a new game starts.
We don't overlap any existing screen elements, unless the upper button list gets really long.
The code is much less crummy (I think, may have made it worse it's hard for me to judge my own work)

If it ends up not being as usable as I'd like, I'll rip out the existing changes and just implement the qol and backend stuff. I think it's worth doing though.

## Changelog
:cl:
add: Expanded heavily on action buttons
add: Adds an action button dropdown that sits just under the normal list in the top left. You can drag new buttons onto it to insert them. Click on it to show its contents, do what you want to do, then click again anywhere to contract it. Alt click it to reset all button positions
add: Action buttons will now remember their position between rounds. So if you really like your flashlight right next to your player for some reason, we support that now
add: When you start to drag an action button, docking ports will appear in places that it can be inserted into. (Outside of just floating somewhere on your screen of course)
del: Removed action button locking, and the associated preference. I'm reasonably sure literally none uses this, but if you do hit me up
qol: Dragging an action button will now give you an outline of its size around your cursor
fix: You can no longer cause the screen to expand by putting an action button on the edge of widescreen, and then resizing to standard.
refactor: Refactors action and button code significantly. lots of little things.
/:cl:

### CODE REVIEW

**Comment by SandPoot** [comment]
Hang on, target appearance is broken, will have to fix.

File: code/_onclick/drag_drop.dm [hunk 1]
```diff
@@ -106,16 +106,30 @@
...
 
 /client/MouseDrag(src_object,atom/over_object,src_location,over_location,src_control,over_control,params)
+	var/list/modifiers = params2list(params)
+	if (LAZYACCESS(modifiers, MIDDLE_CLICK))
+		if (src_object && src_location != over_location)
+			middragtime = world.time
+			middle_drag_atom_ref = WEAKREF(src_object)
+		else
+			middragtime = 0
+			middle_drag_atom_ref = null
```
Comment (deathride58) [inline code/_onclick/drag_drop.dm]: Regression of #15645

File: code/_onclick/drag_drop.dm [hunk 2]
```diff
@@ -106,16 +106,30 @@
...
+	return ..()
 
 /obj/item/proc/onMouseDrag(src_object, over_object, src_location, over_location, params, mob)
 	return
+
+/client/MouseDrop(atom/src_object, atom/over_object, atom/src_location, atom/over_location, src_control, over_control, params)
+	if (IS_WEAKREF_OF(src_object, middle_drag_atom_ref))
+		middragtime = 0
+		middle_drag_atom_ref = null
+	..()
```
Comment (deathride58) [inline code/_onclick/drag_drop.dm]: Regression of #15645

File: code/modules/client/client_defines.dm [hunk 3]
```diff
@@ -134,16 +134,19 @@
...
-	var/mouseObject = null
-	var/mouseControlObject = null
+	///Used in MouseDrag to preserve the last mouse-entered location. Weakref
+	var/datum/weakref/mouse_location_ref = null
+	///Used in MouseDrag to preserve the last mouse-entered object. Weakref
+	var/datum/weakref/mouse_object_ref
+	//Middle-mouse-button click dragtime control for aimbot exploit detection.
+	var/middragtime = 0
+	//Middle-mouse-button clicked object control for aimbot exploit detection. Weakref
+	var/datum/weakref/middle_drag_atom_ref
```
Comment (deathride58) [inline code/modules/client/client_defines.dm]: Regression of #15645

File: code/modules/client/client_procs.dm [hunk 4]
```diff
@@ -881,11 +881,16 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
...
+	var/list/modifiers = params2list(params)
 
-	if(L["drag"])
+	var/dragged = LAZYACCESS(modifiers, DRAG)
+	if(dragged && !LAZYACCESS(modifiers, dragged)) //I don't know what's going on here, but I don't trust it
 		return
 
+	if (object && IS_WEAKREF_OF(object, middle_drag_atom_ref) && LAZYACCESS(modifiers, LEFT_CLICK))
+		ab = max(0, 5 SECONDS-(world.time-middragtime)*0.1)
+
```
Comment (deathride58) [inline code/modules/client/client_procs.dm]: Regression of #15645

File: code/modules/client/client_procs.dm [hunk 5]
```diff
@@ -901,6 +906,9 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 				clicklimiter[ADMINSWARNED_AT] = minute
 
 				msg += " Administrators have been informed."
+				if (ab)
+					log_game("[key_name(src)] is using the middle click aimbot exploit")
+					message_admins("[ADMIN_LOOKUPFLW(usr)] [ADMIN_KICK(usr)] is using the middle click aimbot exploit</span>")
```
Comment (deathride58) [inline code/modules/client/client_procs.dm]: Regression of #15645

**Review by deathride58 — CHANGES_REQUESTED** [review]
Pretty much the only thing that sticks out in review is a partial regression of #15645 (The code I added with that PR is present! But so is the code I removed)

My fix makes TG's older method of checking specifically for middle-click dragging obsolete, as checking specifically for `DRAG` during unrelated mouse-up/mouse-down events (the only cases where `DRAG` gets sent) makes it possible to patch out the aimbot exploit in its entirety

**Review by deathride58 — APPROVED** [review]
There we go!

**Pull request merged**